LEARNER-2530: Messaging changes on Learner Profile

dianakhuang opened the pull request:
## [LEARNER-2530](https://openedx.atlassian.net/browse/LEARNER-2530)

### Description

Add new message to the learner profile, allowing the viewer to dismiss the message. Only show the message once.

![image](https://user-images.githubusercontent.com/5985072/30677038-678d953c-9e57-11e7-8479-77f5dedaca7a.png)

### Acceptance Criteria

- [x] Add a Waffle flag for releasing this feature
- [x] Verify that learners viewing their profile for the first time since the launch of this story are shown the profile introduction message.
- [x] Verify that learners can dismiss this message from the top of their learner profile.
- [x] Verify that learners are only shown this message once and then we auto-dismiss the notice

### Sandbox
- [x] https://dianakhuang.sandbox.edx.org/

### Front-end changes:
- [ ] Accessibility (Check for a11y violations, ensure keyboard accessible, screen reader testing as appropriate)
- [ ] i18n
- [ ] RTL

### Assign Github reviewers (as needed)
- [ ] engineering 
- [ ] product
- [ ] UX
- [ ] documentation

FYI: @edx/learner-mercury
 
### Post-review
- [ ] Rebase and squash commits

andy-armstrong commented:
@dianakhuang @marcotuts @grantgoodmanedX I feel that this message is too dense and hard to read. Could we break it into multiple lines, use bullets, maybe introduce a heading, etc?

![image](https://user-images.githubusercontent.com/5985072/30443070-522b4492-994c-11e7-9a5d-e0c197884251.png)

andy-armstrong commented:
Regarding the dismiss message, if the lines were shorter then I think an "x" on the right would be clearer.

andy-armstrong commented on openedx/features/learner_profile/views/learner_profile.py:
```diff
@@ -42,10 +45,34 @@ def learner_profile(request, username):
+                request.COOKIES.get('dismiss-profile-message', '') != 'True'):
+            PageLevelMessages.register_info_message(
+                request,
+                Text(_('Welcome to the new learner profile! When your profile visibility is set to "Full Profile", \
```
Nit: I believe our convention is to have multiple single line strings, and not break like this.

andy-armstrong commented on openedx/features/learner_profile/views/learner_profile.py:
```diff
@@ -42,10 +45,34 @@ def learner_profile(request, username):
+                request,
+                Text(_('Welcome to the new learner profile! When your profile visibility is set to "Full Profile", \
+                        other edX learners can now see your full name, the certificates you have earned, and when \
+                        you joined the edX community. You can also optionally add links to your social accounts from \
```
This string should not contain "edX" but instead the product name should be used.

andy-armstrong commented on openedx/features/learner_profile/views/learner_profile.py:
```diff
@@ -42,10 +45,34 @@ def learner_profile(request, username):
+        dismiss_message = False
+        if (context['own_profile'] and
+                BOOST_PROFILE_VISIBILITY.is_enabled() and
+                request.COOKIES.get('dismiss-profile-message', '') != 'True'):
```
Use ``not foo`` rather than ``foo != True``.

andy-armstrong replied:
I'd name the cookie for the current state, rather the action that was done. Maybe 'profile-message-viewed' since this is set whether or not the user dismisses it.

dianakhuang replied:
I will change the cookie name, but I don't think I can use `not foo` as the check, because the cookie value will be a string, in this case `'True'`.

andy-armstrong replied:
Good point. I thought there was a helper function for handling boolean parameters like this, but I can't find one.

andy-armstrong commented on openedx/features/learner_profile/views/learner_profile.py:
```diff
@@ -42,10 +45,34 @@ def learner_profile(request, username):
+                    )
+                )
+            )
+            dismiss_message = True
```
I think you could just set the cookie here and not have to use a variable.

dianakhuang replied:
The cookie is set on the response, not the request. I need to generate the response before I can set the cookie. It causes all these messy dependencies where I don't think I can generate the response before I register the page-level message or else the message won't be there.

andy-armstrong replied:
Ah, good point. Makes sense.

andy-armstrong commented on openedx/features/learner_profile/__init__.py:
```diff
@@ -11,3 +11,7 @@
+
+# Waffle flag for increasing the visibility of the profile page, including all messaging necessary to show it.
+# TODO: LEARNER-2554: 09/2017: Remove flag once message is no longer needed.
+BOOST_PROFILE_VISIBILITY = WaffleFlag(WAFFLE_FLAG_NAMESPACE, 'boost_visibility')
```
I don't see the change to make the header obey this Waffle flag. Are you planning to add that to the other PR once this merges?

dianakhuang replied:
We discussed this during standup. It's a pain to try to wire up the header changes to a waffle flag since it touches so many files. We're going to merge this in first and then merge the header changes when they're ready.

andy-armstrong replied:
In that case I'd rename the flag to something like SHOW_NEW_PROFILE_MESSAGE. The comment and name as written don't fit with how it is going to be used.

dianakhuang commented:
I have addressed most of the code review feedback with the exception of fixing the issues with the message text. I'll handle those issues when we get the finalized copy.

andy-armstrong commented:
Thanks for addressing my feedback, @dianakhuang. The code all looks great, and I'm looking forward to seeing where we land on the message design.

grantgoodmanedX commented:
Here's a shorter message:
Welcome to the new learner profile page. Your full profile now displays more information to other learners. You can instead choose to display a limited profile. <Learn more| http://edx.readthedocs.io/projects/edx-guide-for-students/en/latest/SFD_dashboard_profile_SectionHead.html#adding-profile-information>

dianakhuang commented:
@marcotuts @andy-armstrong I've updated the message text in the code and on the sandbox.

andy-armstrong commented on openedx/features/learner_profile/views/learner_profile.py:
```diff
@@ -42,10 +45,39 @@ def learner_profile(request, username):
+            message_text = Text(_('Welcome to the new learner profile page. Your full profile now displays more '
+                                  'information to other learners. You can instead choose to display a limited '
+                                  'profile. {learn_more_link_start}Learn more{learn_more_link_end}')).format(
+                learn_more_link_start=HTML('<a href="http://edx.readthedocs.io/projects/open-edx-learner-guide/en/'
```
I just realized that this needs to go through the help lookup system, so that it correctly picks up Open edX vs edX vs community documentation.

andy-armstrong commented:
@dianakhuang @marcotuts @HarryRein I've updated Diana's sandbox with my styling improvements:

https://dianakhuang.sandbox.edx.org/u/verified

I'm having difficulty removing the white area between the two message areas. If one of you is able to track that down that would be awesome.

![image](https://user-images.githubusercontent.com/5985072/30658402-9013f3a6-9e08-11e7-8f3f-e2930dad7bee.png)

I've just realized that I've probably also broken the dismiss button, so I'll look at that when I get back. I generalized it so that any page message can have a dismiss button but forgot to update the JS to match.

dianakhuang commented:
jenkins run js

andy-armstrong commented:
@dianakhuang @marcotuts I've implemented the styling to my satisfaction, and fixed the dismiss button that I broke. If this is green it should be ready to merge in the morning. @marcotuts check out the screen shot in the PR description above to see if it matches what you had in mind.

andy-armstrong commented:
jenkins run python

marcotuts commented on lms/static/sass/bootstrap/_components.scss:
```diff
@@ -13,7 +13,13 @@
+
   .icon-alert {
-    margin-right: $baseline / 4;
+    margin-right: $baseline / 2;
```
NIT: $baseline suggested, helps balance out spacing on left and right of icon.

marcotuts commented on lms/static/sass/elements/_banners.scss:
```diff
@@ -61,19 +61,31 @@ $full-width-banner-margin: 20px;
+            @include margin-left($baseline / 2);
         }
 
         &.alert-info {
```
Hmm... NIT as well, but line-height: initial? The text lines are a little tight here on the message.
